File: src-tauri/src/app/modules/core/list.rs
use crate::app::modules::core::ops::manager::ModManager;
use crate::app::modules::io::use_io;
use clap::Args;

#[derive(Args)]
pub struct List {}

impl List {
    pub async fn run() -> Result<(), String> {
        let io = use_io();

        let mut manager = ModManager::new();
        manager.load().await;
        manager.with_default_providers();

        let mods = manager.manifest_service.manifest.mods_as_entries();

        if mods.is_empty() {
            io.info("No mods installed.");
            return Ok(());
        }

        io.info("Installed mods:");

        for entry in mods {
            io.print(&format!(
                " - {}:{} ({})",
                entry.provider.to_string().to_lowercase(),
                entry.slug,
                entry.version.as_str()
            ));
        }

        Ok(())
    }
}

File: packages/core/src/operations/list.ts
import type {ModManager} from "./mod-manager.js";
import {versionSpecToString} from "../models/manifest.js";

export class List {
  static async run(manager: ModManager): Promise<void> {
    const mods = manager.manifestEntries("mod");
    const datapacks = manager.manifestEntries("datapack");

    if (mods.length === 0 && datapacks.length === 0) {
      manager.io.info("No mods or datapacks installed.");
      return;
    }

    if (mods.length > 0) {
      manager.io.info("Installed mods:");
      for (const entry of mods) {
        const suffix = entry.disabled ? " [disabled]" : "";
        manager.io.print(
          ` - ${entry.provider}:${entry.slug} (${versionSpecToString(entry.version)})${suffix}`,
        );
      }
    }

    if (datapacks.length > 0) {
      manager.io.info("Installed datapacks:");
      for (const entry of datapacks) {
        const suffix = entry.disabled ? " [disabled]" : "";
        manager.io.print(
          ` - ${entry.provider}:${entry.slug} (${versionSpecToString(entry.version)})${suffix}`,
        );
      }
    }
  }
}
